R/060_utils.R

```diff
@@ -0,0 +1,348 @@
+#'
+
+reduce_mutated <- function(df, form_list, format) {
+  current_quosure <- form_list[[1]]
+  var_name_sym <- rlang::sym(form_list[[2]])
+
+  df %>%
+    dplyr::mutate(!!var_name_sym := !!current_quosure)
+}
+
+#' Give quosure a name
+#'
+#' This is an internal function that adds a prefixed name to a quosure so that variables added to a data frame using this quosure have a predictable name.
+#' @param x quosure
+#' @param prefix prefix to be added in name
+#'
+
+append_name_to_quosure <- function(x, prefix = "grp_") {
+  list(
+    x,
+    paste0(
+      prefix,
+      x %>% rlang::as_label() %>% make.names() %>%
+        stringr::str_replace_all("\\.+", ".") %>% stringr::str_remove_all("(\\.$)|(^X\\.)") %>%
+        stringr::str_replace_all("\\.", "_") %>%
+        ifelse(stringr::str_sub(., start = 1, 1) %in% as.character(0:9), paste0("x", .), .)
+    )
+  )
+}
+
+#' Add a variable to a tidyxl data frame if that variable is missing
+#'
+#' This is an internal function that takes a tidyxl data frame and a variable name to be added if they are missing.
+#' @param sheet sheet nominated for tidying
+#' @param var name of the var to be added as a character string.
+#'
+add_variable_if_missing <- function(sheet, var) {
+  if (!var %in% names(sheet)) {
+    var_sym <- rlang::sym(var)
+
+    sheet <- sheet %>% dplyr::mutate(!!var_sym := NA_character_)
+  }
+
+  sheet
+}
+
+#' Create a filtering variable based on a excel string range.
+#'
+#' This is an internal function that create a variabled prefixed "flt" based on whether cells are within a range specified by an excel string range.
+#' @param table_range a table_range given by string. For example, "A1:B2"
+#' @param sheet a tidyxl data frame.
+
+string_range_to_filter_vars <- function(sheet, table_range) {
+  cell_ref_df <- tibble::as_tibble(cellranger::as.cell_limits(table_range))
+
+  table_range_df <-
+    cell_ref_df[, 1:2] %>%
+    purrr::set_names(c("min", "max")) %>%
+    dplyr::mutate(dimension = c("row", "col")) %>%
+    tidyr::gather(key, value, -dimension) %>%
+    tidyr::unite(label, key, dimension, sep = "_") %>%
+    tidyr::spread(label, value)
+
+  string_filter_name <- rlang::sym(table_range %>% stringr::str_remove("\\:") %>% paste0("flt_", .))
+
+  data_sheet <-
+    sheet %>%
+    dplyr::mutate(!!string_filter_name :=
+      row >= table_range_df$min_row[1] &
+        row <= table_range_df$max_row[1] &
+        col >= table_range_df$min_col[1] &
+        col <= table_range_df$max_col[1])
+}
+
+#' Get the corner references of a tidyxl data frame.
+#'
+#' This is an internal function that takes a tidyxl data frame a returns a dataframe indicating the maximum and minimum corners.
+#'
+#' @param sheet a tidyxl data frame.
+
+get_corner_cell_refs <- function(sheet) {
+
+  # Automatic producedure
+  ref_df <-
+    sheet %>%
+    dplyr::summarise(
+      min_row = min(row), max_row = max(row),
+      min_col = min(col), max_col = max(col)
+    )
+  return(ref_df)
+}
```

Comment:
I'd prefer this to return a simple named list rather than a data frame.  The only difference is that the return value wouldn't have the `tbl_df` or `data.frame` classes, and I don't think it would break any other code to just return a list.

```r
get_corner_cell_refs <- function(sheet) {
    list(min_row = min(sheet$row),
         max_row = max(sheet$row),
         min_col = min(sheet$col),
         max_col = max(sheet$col)
         )
}
```

Reply:
Sounds good to me.

Reply:
Addressed in 7fb79db9990b511bc69fbdd85b32169989c33e0e.